src/lightbox/LightboxHeader.js

```diff
@@ -54,7 +55,12 @@ export default class LightboxHeader extends PureComponent<Props> {
 
     return (
       <View style={styles.wrapper}>
-        <UserAvatarWithPresence size={36} avatarUrl={avatarUrl} />
+        <UserAvatarWithPresence
+          size={36}
+          // In the next commit, `UserAvatarWithPresence` will take an
+          // AvatarURL instance for `avatarUrl`.
+          avatarUrl={avatarUrl.get(36).toString()}
```

Comment:
> Can we do an NFC commit before this one that turns the latter into thin wrappers around the former?

Hmm, I'm probably missing something, so I'd better check. A commit that turns old helper functions into thin wrappers around new code doesn't sound NFC; is that right? In that commit, the old helper functions' implementations would go from being used to not being used.

Reply:
> > Can we do an NFC commit before this one that turns the latter into thin wrappers around the former?
> 
> Hmm, I'm probably missing something, so I'd better check. A commit that turns old helper functions into thin wrappers around new code doesn't sound NFC; is that right? In a commit like that, the old helper functions' old implementations would go from being used to not being used.

Right, so such a commit would stop using the old implementation and start using the new implementation.

Whether that's NFC depends on whether there are differences in behavior between the old and the new implementations. If the new implementation is purely a new way of organizing the same ultimate logic as the old implementation has, so there's no change in behavior, then that commit is NFC.

I think this refactor is supposed to be at least *mostly* like that, right? At the end there's the commit that adds the fallback behavior -- but before that, IIUC it's not intended to change much behavior, if any. Is there some behavior it does intentionally change? Or, is there some behavior that it's hard to avoid having it change?

And another way to put that question is: if we do, before this commit, a change that looks something like:
```diff
 export const getAvatarFromUser = (user: UserOrBot, realm: URL, size?: number): string =>
-  getAvatarUrl(user.avatar_url, user.email, realm, size);
+  AvatarURL.fromUserOrBotData({ rawAvatarUrl: user.avatar_url, email: user.email, realm }).get(size).toString();
 
 export const getAvatarFromMessage = (
   message: Message | Outbox,
   realm: URL,
   size?: number,
-): string => getAvatarUrl(message.avatar_url, message.sender_email, realm, size);
+): string => {
+  const avatarUrl = AvatarURL.fromUserOrBotData({ rawAvatarUrl: message.avatar_url, email: message.sender_email, realm });
+  return avatarUrl.get(size).toString();
+};
```
then would that be NFC? And if not, why/how not?

If the answer is "yes", then arranging the series so that it contains that commit would be a nice way of demonstrating that. And if the answer is "no", then the detailed answer will be interesting for helping understand the branch... and rearranging things so that there is such a commit, if that's possible, would help make that detailed answer concrete, and help isolate the things that do change (in their own, focused commits) so that they're easier to see and to think about.

Reply:
Ah OK, got it, I think—I hadn't been thinking of calling `.toString()` and returning a string, as you do in that example; that does make it easier to focus on how it might be NFC.

You've pointed to two exported functions, `getAvatarFromUser` and `getAvatarFromMessage`, being used to compute the actual URL from the pattern the server provides in `avatar_url`. Awkwardly, there's a third: `getAvatarUrl`. I have a hunch that it began its life as a helper function for those other two, but that `UserAvatarWithPresence` saw it as an attractive option because it's sometimes passed info about a user, and sometimes about a message. I suspect we can make the same kind of change to `getAvatarUrl` as you point out for those other two, though.

I find a few things that could be described as functional changes; maybe they're pretty minor (?):

1.

In the old code, there's a tiny bug stemming from an assumption that no `avatar_url` that's a string will be a Gravatar URL. There are in fact two cases where we'll get a string that represents a Gravatar URL. These are noted in comments on `AvatarURL.fromUserOrBotData`.
  - If we don't announce `client_gravatar`. At that point in the branch, we don't announce it in `getMessages`. The server doesn't know that we can compute the Gravatar URL ourselves, so it just gives it to us.
  - If the server doesn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set. I think this is because people will want the Gravatar URL associated with their real email address to be used, without the real email address ever appearing client-side. The server knows that we can't give that proper Gravatar URL if we don't have the real email address, so it just gives us the proper Gravatar URL.

Reading past the first conditional in `getAvatarUrl`, we appear to assume that the `avatar_url` string will correspond to a relative or absolute URL for an uploaded avatar. This code is mostly harmless to a Gravatar URL string that has snuck through for one of the above reasons. It's an absolute URL, so `realm` in `new URL(avatarUrl, realm)` will be ignored. And if `size` is greater than 100, it'll be run through a regex, but that's a no-op because Gravatar URLs don't have `.png` in them.

I say it's "mostly harmless" because a Gravatar URL string sneaking through in these cases won't ever be set to the desired or Zulip-default size. The way to set a Gravatar URL to a certain size is to pass a query param `s`, like `s=100`. And that's just not done in these cases because the code isn't written to be aware that it might be working with a Gravatar URL.

The new code handles these cases by recognizing that we're dealing with Gravatar URLs, and they will be sized appropriately.

2.

The old code uses a default size of 80. The new code doesn't use a default size, but treats a not-provided size as saying "I don't really mind what size I get." For Gravatar URLs, this means not passing the `s` query param, and leaving it up to Gravatar to pick a size. For uploaded avatars, it means using the non-`-medium.png` version, which is what we also do when we pass a size that's 100 or less.

Reply:
> You've pointed to two exported functions, `getAvatarFromUser` and `getAvatarFromMessage` [...] I suspect we can make the same kind of change to `getAvatarUrl` as you point out for those other two, though.

Ah, good catch, yes. I think that means a switchover commit could just edit `getAvatarUrl` instead of those two callers of it.

---

> * If the server doesn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set. When that's set, real email addresses don't get sent to the client. But I think real email addresses are still supposed to determine Gravatar URLs, since I might have told Gravatar what my email address is, and what image to show for me. The server can't expect the client to compute the Gravatar URL for the real email address, since the server isn't sending real email addresses. So they just give the client the Gravatar URL for the real email address.

Yep, this is correct -- when emails are hidden, the server still sends Gravatar URLs based on the real emails. (I'm not sure this is documented anywhere; I don't see it [in the user-facing docs](https://zulip.com/help/restrict-visibility-of-email-addresses). But it's been confirmed by Tim in chat threads.)

---

> 2. The old code uses a default size of 80. The new code doesn't use a default size, but treats a not-provided size as saying "I don't really mind what size I get." For Gravatar URLs, this means not passing the s query param, and leaving it up to Gravatar to pick a size. For uploaded avatars, it means using the non-`-medium.png` version, which is what we also do when we pass a size that's 100 or less.

Ah, that summary is helpful, thanks.

Looking up what Gravatar does in that case... [its default is 80](https://en.gravatar.com/site/implement/images/)! I'd guess that's where our code's default of 80 came from, now that I see that. :slightly_smiling_face: So I think that means this doesn't actually represent a change in behavior, once that small fact from the Gravatar API (which is the API we're effectively using when constructing these Gravatar URLs) is taken into account.

---

> 1. In the old code, there's a tiny bug stemming from an assumption that, if `avatar_url` is a string, it won't be a Gravatar URL. There are in fact two cases where we'll get a string that represents a Gravatar URL.

Mmm, yeah. And in master, or at the beginning of the branch, there's a third: the "double-processing" bug means we produce a Gravatar URL string prematurely, and then pass it through this avatar-URL logic again.


> I say it's "mostly harmless" because there is some behavior we probably don't intend: a Gravatar URL string sneaking through in these cases won't ever be set to the desired [size]. The way to set a Gravatar URL to a certain size is to pass a query param `s`, like `s=100`. And that's just not done in these cases because the code isn't written to be aware that it might be working with a Gravatar URL.

Ah indeed. I think I learned about some of this shortly after I wrote the comment above, in fact: while writing up and repro'ing #4305 (which grew from this comment elsewhere on the present PR: https://github.com/zulip/zulip-mobile/pull/4230#discussion_r521685284 ). It's one reason the symptoms of that issue are different between Gravatars and uploaded avatars; though both cases are buggy, because when we do specify a size it's always too small (that's the core of #4305.)

I think this point gives us an example of this idea I described above, then:

> And if the answer is "no", then the detailed answer will be interesting for helping understand the branch... and rearranging things so that there is such a commit, if that's possible, would help make that detailed answer concrete, and help isolate the things that do change (in their own, focused commits) so that they're easier to see and to think about.

Specifically: how about a commit early in the branch that does the equivalent of this:
```js
      if (tryParseUrl(rawAvatarUrl)?.origin === GravatarURL.ORIGIN) {
        const hashMatch = /[0-9a-fA-F]{32}$/.exec(new URL(rawAvatarUrl).pathname);
        if (hashMatch === null) {
          const msg = 'Unexpected Gravatar URL shape from server.';
          logging.error(msg, { value: rawAvatarUrl });
          throw new Error(msg);
        }
        return GravatarURL.validateAndConstructInstance({ email, hash: hashMatch[0] });
      }
```
but within the string-heavy, no-classes idiom of the existing code? I think that'd be a pretty small commit. The hard part here was in your sorting out the exact set of different situations that can occur. That commit would be a way of expressing this piece of what you learned; its commit message would probably reuse a chunk of the text from your comment here in this subthread.

---

Then with that change to fix the Gravatar-URL-string case, plus the "double-processing" fix... plus either a commit to adjust the old implementation's default-size behavior to match the new one's, or tweaking the new one's to be more like the old one's... I think at that point a switchover commit like the one I described would indeed be NFC. Does that sound right?

Reply:
> Does that sound right?

Yes, it does! Thanks for these helpful pointers.